### DISCUSSION TITLE
Input file for turbine power output simulations

### DISCUSSION THREAD

**Discussion by vayuai** [Exawind/amr-wind#489] [Q&A]
Hi, I'm a pretty new amr-wind user and have been trying to use the package to run simulations to study the effect of wake on turbine power output.
My first attempt was to use a very basic method, simply using a uniform flow simulation (no ABL or FAST) to get some understanding of what the results might look like.
The problem I ran into was figuring out what the input file should look like, as the documentation can be fairly nebulous at times (and some of the input file examples in there were throwing errors when I tried to run them).
After piecing together stuff from the documentation and some of the test code I managed to come up with an input file that seems like it should have worked. But looking at the plotfiles that resulted from it, it doesn't seem like there is any variable in there that looks like power output (and also the range for the velocity variable suggest to me that wake hasn't been simulated properly as there should have been a wider spread for sure).

Github won't let me attach a .abl file so here are the contents of the input file:
```
geometry.prob_lo        =   0.0 0.0 -50.0
geometry.prob_hi        =   2000.0  2000.0  350.0
geometry.is_periodic    =   0   0   0   # Periodicity x y z (0/1)
amr.n_cell                                  = 200 200 40
amr.max_level                               = 0
time.stop_time                              = 500.0
time.max_step                               = -1
time.fixed_dt                               = 0.5
time.cfl                                    = 0.95
time.plot_interval                          = 10
time.checkpoint_interval                    = -1
io.KE_int                                   = 0
io.check_file                               = "chk"
io.plot_file                                = "plt"
io.outputs = actuator_src_term
incflo.physics                              = FreeStream Actuator
ConstValue.density.value = 1.0
ConstValue.velocity.value = 5.0 0.0 0.0
incflo.verbose                              = 0
incflo.initial_iterations                   = 3
incflo.do_initial_proj                      = true
incflo.use_godunov                          = 1
incflow.godunov_type                        = "ppm_nolim"
transport.viscosity                         = 1.0e-5
transport.laminar_prandtl                   = 0.7
transport.turbulent_prandtl                 = 0.3333
turbulence.model                            = Smagorinsky
Smagorinsky_coeffs.Cs = 0.16
ICNS.source_terms                           = ActuatorForcing
xlo.type = "mass_inflow"
xlo.density = 1.0
xlo.velocity = 5.0 0.0 0.0
xhi.type = "pressure_outflow"
ylo.type =   "slip_wall"
yhi.type =   "slip_wall"
zlo.type =   "slip_wall"
zhi.type =   "slip_wall"
nodal_proj.verbose = 0
nodal_proj.mg_rtol = 1.0e-6
nodal_proj.mg_atol = 1.0e-12
mac_proj.mg_rtol = 1.0e-6
mac_proj.mg_atol = 1.0e-12
Actuator.labels = WTG01 WTG02
Actuator.type = UniformCtDisk
Actuator.UniformCtDisk.rotor_diameter = 126.0
Actuator.WTG01.base_position = 748.0 0.0 0.
Actuator.WTG02.base_position = 1252.0 -151.2 0.
Actuator.WTG01.yaw = 250.0
Actuator.WTG02.yaw = 270.0
Actuator.WTG01.sample_yaw = 250.0
Actuator.WTG02.sample_yaw = 270.0
Actuator.UniformCtDisk.hub_height = 90.0
Actuator.UniformCtDisk.num_force_points = 5
Actuator.UniformCtDisk.thrust_coeff = 0.0 0.7 1.2
Actuator.UniformCtDisk.wind_speed = 0.0 10.0 12.0
Actuator.UniformCtDisk.epsilon = 10.0
Actuator.UniformCtDisk.density = 1.225
Actuator.UniformCtDisk.diameters_to_sample = 1.0
Actuator.UniformCtDisk.num_vel_points_r = 3
Actuator.UniformCtDisk.num_vel_points_t = 3
tagging.labels = refine0
tagging.refine0.type = GeometryRefinement
tagging.refine0.shapes = c0 c1
tagging.refine0.c0.type = cylinder
tagging.refine0.c0.start = -256.0 0 0
tagging.refine0.c0.end = 2600.0 0 0
tagging.refine0.c0.outer_radius = 140.0 
tagging.refine0.c1.type = cylinder
tagging.refine0.c1.start = -189.0 0 0
tagging.refine0.c1.end = 2646.0 0 0
tagging.refine0.c1.outer_radius = 70.0
```

I would greatly appreciate any help/advice that can elucidate what is wrong with this input file.

**Comment by sayerhs** [comment]
Tagging @psakievich @lawrenceccheung

**Comment by psakievich** [comment]
@vayuai I apologize about the documentation. You are correct. There isn't much there for this particular feature yet.  It is still a work in progress.  In terms of the results you are also correct that there is no power estimate from this model yet.  We don't have a Cp variable added, but this is something we can add pretty easily.  For the V&V efforts we've mainly been looking at velocity deficit and wake profiles.  In terms of the wake structure would you mind sharing some plots of what you are seeing?

**Comment by lawrenceccheung** [comment]
Hi @vayuai, 
I noticed a couple of things about the input file shown above.  The first is regarding the turbine placement:
```
Actuator.WTG01.base_position = 748.0 0.0 0.
Actuator.WTG02.base_position = 1252.0 -151.2 0.
```
Given that the domain runs from 0 to 2000 in both x and y, that means `WTG01` will on the edge of one boundary, and `WTG02` will be outside the domain:
![image](https://user-images.githubusercontent.com/15526007/137784383-ec56e330-a951-4a7d-b148-3f75871e57b1.png)
Both turbines should be closer to the middle of the domain to get better wakes.

`amr.max_level` is also set to 0, meaning the refinement zones in the `tagging` section won't be applied.  Finally, `Actuator.UniformCtDisk.density` is set to `1.225`, but both `xlo.density` and `ConstValue.density.value` are 1.0, so that might be a point of inconsistency.

Hope that helps!  Let me know if you have any other questions.

Lawrence

**Comment by vayuai** [comment]
@lawrenceccheung thank you for all the comments. I will certainly fix all these inputs for the next trial.
@psakievich I can share a plotfile but I think the problem with the turbine locations being at the edge of the grid made the results useless (and led to the range in the velocity across the field being essentially 0). I would need to rerun with the tweaks suggested above and see what that looks like then.
Regarding the Cp variable, is there any way of generating power without changes to the current code? I do know that amr-wind is supposed to be able to interface with OpenFAST but I haven't managed to find any good examples and I also don't know if this would be a way of solving the problem at all (but I imagine OpenFAST does produce power output).
I would really appreciate it if you could provide any guidance for how I might go about running a simulation that outputs Cp.

**Reply by psakievich** [reply]
yes no need to share a plotfile given the changes.

**Comment by lawrenceccheung** [comment]
hi @vayuai,

Theoretically speaking, using the current `UnifCtDisk` model that's in there now, you could manually sample the points on the rotor disk and calculate the turbine power that way.  However, that's not something I recommend doing unless you specifically wanted the power estimate from this precise turbine model.   Also, the uniform Ct disk model is currently an idealized model, and for high Ct values (Ct>0.8 or so), the behavior gets to be unphysical.  But I'll talk to @psakievich and see about including the Cp value in the calculation, since it's a handy value to have nonetheless.

The alternative, as you mentioned, is to go with the OpenFAST model (which does output turbine power).  One thing that might help in this regard is a user front-end tool to AMR-Wind that I've been working on: https://github.com/lawrenceccheung/amrwind-frontend.  This tool will help set up the run and put in the proper turbines and refinement regions, and might help avoid any common setup mistakes in the input file.  There's also a handy tutorial on setting up a case with the NREL 5MW turbine using OpenFAST that can be found here: https://github.com/lawrenceccheung/amrwind-frontend/blob/main/docs/tutorial1/README.md.

Cheers,

Lawrence
